### PR TITLE
Issue 17698 webdav casing issues

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
Our implementation of WebDav currently does not support Different Casing. Everything uploaded via WebDav ends-up in lowercase. Another PR fixed that small problem.  But.. while testing that fix it was discovered that our implementation wasn't fully prepared to safely handle the rename operation. 

When we do rename a file and change the casing. The Operation is handled by the library as a 
`Move command` which takes a source and target. 
When we say rename the resource `/webdav/live/1/demo.dotcms.com/images/black.png`  to `/webdav/live/1/demo.dotcms.com/images/BLACK.png` 
we're instructing WebDav to move a file where the source and the destination are the same.
By default, the library takes the precaution to delete the destination resource to ensure the move is successful.  But in this case, the source and the destination are interpreted as the same resource.

meaning that 
 `/webdav/live/1/demo.dotcms.com/images/black.png` 
and 
`/webdav/live/1/demo.dotcms.com/images/BLACK.png` 

Are the same file.
This PR solves that problem.

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        FileAssetDataGen fileAssetDataGen = new FileAssetDataGen(child, file);
+        fileAssetDataGen.nextPersisted();
+        final String path = String.format("http://localhost:8080/webdav/live/1/%s/%s/%s/%s",host.getName(),parent.getName(),child.getName(),file.getName());
+        final String fileResourcePath1 = upperCaseRandom(path,8);
+        final String fileResourcePath2 = upperCaseRandom(path,10);
+        Assert.assertNotEquals(fileResourcePath1,fileResourcePath2);
+        Assert.assertTrue(helper.isSameResourceURL(fileResourcePath1,fileResourcePath2, file.getName()));
+    }
+
+    private static String upperCaseRandom(final String input, final int n) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java]: Codacy Issue found: [Avoid variables with short names like n](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4792071)

File: dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        Assert.assertTrue(folderResource instanceof FolderResource);
+        final String fileResourcePath = upperCaseRandom(String.format("http://localhost:8080/webdav/live/1/%s/%s/%s/%s",host.getName(),parent.getName(),child.getName(),file.getName()),8);
+        final Resource fileResource = helper.getResourceFromURL(fileResourcePath);
+        Assert.assertNotNull(fileResource);
+        Assert.assertTrue(fileResource instanceof FileResource);
+    }
+
+    @Test
+    public void Test_Get_Folder_Resource_For_Non_Existing_Path() throws IOException, DotDataException, DotSecurityException {
+        String path = "http://localhost:8080/webdav/live/1/demo.dotcms.com/images/black.png";
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java]: Codacy Issue found: [Local variable 'path' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4792071)

File: dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    @Test
+    public void Test_Same_Resource_For_Paths_With_Different_Casing() throws IOException, DotDataException, DotSecurityException {
+        final SiteDataGen siteDataGen = new SiteDataGen();
+        final FolderDataGen folderDataGen = new FolderDataGen();
+        final Host host = siteDataGen.nextPersisted();
+        final Folder parent = folderDataGen.site(host).nextPersisted();
+        final Folder child = folderDataGen.parent(parent).nextPersisted();
+        java.io.File file = java.io.File.createTempFile("texto", ".txt");
+        FileUtil.write(file, "helloworld");
+        FileAssetDataGen fileAssetDataGen = new FileAssetDataGen(child, file);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java]: Codacy Issue found: [Local variable 'fileAssetDataGen' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4792071)

File: dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @Test
+    public void Test_Same_Resource_For_Paths_With_Different_Casing() throws IOException, DotDataException, DotSecurityException {
+        final SiteDataGen siteDataGen = new SiteDataGen();
+        final FolderDataGen folderDataGen = new FolderDataGen();
+        final Host host = siteDataGen.nextPersisted();
+        final Folder parent = folderDataGen.site(host).nextPersisted();
+        final Folder child = folderDataGen.parent(parent).nextPersisted();
+        java.io.File file = java.io.File.createTempFile("texto", ".txt");
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/webdav/DotWebdavHelperTest.java]: Codacy Issue found: [Local variable 'file' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4792071)

File: dotCMS/src/main/java/com/dotmarketing/webdav/ResourceFactorytImpl.java [hunk 5]
```diff
@@ -43,26 +43,30 @@ public ResourceFactorytImpl() {
 	 */
     @WrapInTransaction
 	public Resource getResource(String davHost, String url) {
+		return getResource(davHost, url, dotDavHelper, hostAPI);
+	}
+
+	public static Resource getResource(final String davHost, String url, final DotWebdavHelper dotDavHelper, final HostAPI hostAPI) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/webdav/ResourceFactorytImpl.java]: Codacy Issue found: [Avoid reassigning parameters such as 'url'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4792071)

**Review by jgambarios — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java [hunk 6]
```diff
@@ -1069,30 +1090,39 @@ public void move(String fromPath, String toPath, User user,boolean autoPublish)t
...
-						newversion.setLanguageId(defaultLang);
-						newversion = conAPI.checkin(newversion, user, false);
+						newVersion.setBinary(FileAssetAPI.BINARY_FIELD, tmp);
+						newVersion.setLanguageId(defaultLang);
+						newVersion = conAPI.checkin(newVersion, user, false);
 						if(autoPublish) {
-							conAPI.publish(newversion, user, false);
+							conAPI.publish(newVersion, user, false);
+						}
+						if(sameSourceAndTarget){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java]: good

File: dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java [hunk 7]
```diff
@@ -1109,32 +1139,44 @@ public void move(String fromPath, String toPath, User user,boolean autoPublish)t
...
-						removeObject(toPath, user);
-						fc.removeFolder(folder, idapi.find(folder));
-					}catch (Exception e) {
-						Logger.debug(this, "Unable to delete toPath " + toPath);
+					//need to verify the source and destination are not the same because we could be renaming the folder to be the same but with different casing.
+					if (!sourceAndDestinationAreTheSame) {
+						try {
+							// Folder must end with "/", otherwise we get the parent folder
+							String folderToPath = getPath(toPath);
+							if (!folderToPath.endsWith("/")) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java]: we have contants on StringUtils for it

File: dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java [hunk 8]
```diff
@@ -1277,7 +1319,125 @@ public void removeObject(String uri, User user) throws IOException, DotDataExcep
...
+				host = sourceParts.getHost();
+			}
+
+			if (UtilMethods.isSet(sourceParts.getPath())) {
+				path = sourceParts.getPath();
+			}
+
+			if (UtilMethods.isSet(sourceParts.getResource())) {
+				if (UtilMethods.isSet(path)) {
+					if (!path.endsWith("/")) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/webdav/DotWebdavHelper.java]: StringUtils

**Review by jdotcms — APPROVED** [review]

**Pull request merged**